### PR TITLE
MDEV-14474 I_S.check_constraint

### PR DESCRIPTION
**Pull request by an3l**

### CODE REVIEW

File: sql/sql_show.cc [hunk 1]
```diff
@@ -6883,6 +6883,68 @@ store_constraints(THD *thd, TABLE *table, const LEX_CSTRING *db_name,
...
+        table->field[4]->store(table_name->str, table_name->length, cs);
+        field->check_constraint->print(&str);
+        table->field[5]->store(str.ptr(), strlen(str.ptr()), cs);
+        schema_table_store_record(thd, table);
+      }
+    }
+
+    if (share->table_check_constraints)
+    {
+      for (uint i=share->field_check_constraints;i<share->table_check_constraints;i++)
```
Comment (svoj) [inline sql/sql_show.cc]: I wonder what this string was supposed to do?

File: sql/sql_show.cc [hunk 2]
```diff
@@ -6883,6 +6883,68 @@ store_constraints(THD *thd, TABLE *table, const LEX_CSTRING *db_name,
...
+      push_warning(thd, Sql_condition::WARN_LEVEL_WARN,
+                   thd->get_stmt_da()->sql_errno(),
+                   thd->get_stmt_da()->message());
+    thd->clear_error();
+    DBUG_RETURN(0);
+  }
+  else if (!tables->view)
+  {
+    TABLE_SHARE *share= tables->table->s;
+    CHARSET_INFO *cs= system_charset_info;
```
Comment (svoj) [inline sql/sql_show.cc]: Why not to use `system_charset_info` directly?

File: sql/sql_show.cc [hunk 3]
```diff
@@ -6883,6 +6883,68 @@ store_constraints(THD *thd, TABLE *table, const LEX_CSTRING *db_name,
...
+    Field **ptr, *field;
+    for (ptr=tables->table->field; (field=*ptr); ptr++)
+    {
+
+      if (field->invisible > INVISIBLE_USER)
+        continue;
+
+      if (field->check_constraint)
+      {
+        StringBuffer<MAX_FIELD_WIDTH> str(&my_charset_utf8mb4_general_ci);
```
Comment (svoj) [inline sql/sql_show.cc]: Why not `str(system_charset_info)`?

File: sql/sql_show.cc [hunk 4]
```diff
@@ -6883,6 +6883,68 @@ store_constraints(THD *thd, TABLE *table, const LEX_CSTRING *db_name,
...
+
+        restore_record(table, s->default_values);
+        table->field[0]->store(STRING_WITH_LEN("def"),cs);
+        table->field[1]->store(db_name->str, db_name->length, cs);
+        table->field[2]->store(STRING_WITH_LEN("NULL"),cs);
+        table->field[3]->store(db_name->str, db_name->length, cs);
+        table->field[4]->store(table_name->str, table_name->length, cs);
+        field->check_constraint->print(&str);
+        table->field[5]->store(str.ptr(), strlen(str.ptr()), cs);
```
Comment (svoj) [inline sql/sql_show.cc]: Why not `str.length()`?

File: sql/sql_show.cc [hunk 5]
```diff
@@ -6883,6 +6883,68 @@ store_constraints(THD *thd, TABLE *table, const LEX_CSTRING *db_name,
   return schema_table_store_record(thd, table);
 }
 
+static int get_check_constraints_record(THD *thd, TABLE_LIST *tables,
+					TABLE *table, bool res,
```
Comment (svoj) [inline sql/sql_show.cc]: Please don't use tabs for indentation, use spaces instead.

File: sql/sql_show.cc (hunk 1 above)
Comment (svoj) [reply sql/sql_show.cc]: Ok, apparently it seem to be correct.

File: mysql-test/t/MDEV14474-I_S_check_constraint.test [hunk 6]
```diff
@@ -0,0 +1,57 @@
+--echo #
+--echo # Task MDEV14474: Create INFORMATION_SCHEMA.CHECK_CONSTRAINTS
```
Comment (cvicentiu) [inline mysql-test/t/MDEV14474-I_S_check_constraint.test]: Usually we don't write it as
`Task MDEV14474: Create INFORMATION_SCHEMA.CHECK_CONSTRAINTS`

Instead we write it like so:
`MDEV-14474: Create INFORMATION_SCHEMA.CHECK_CONSTRAINTS`

**Comment by svoj** [comment]
I'd suggest to follow current naming convention for I_S tests and have information_schema prefix.

Also please run funcs_1 and funcs_2 suites, I tend to remember there were some sensitive tests.

File: sql/sql_show.cc [hunk 7]
```diff
@@ -6883,6 +6883,67 @@ store_constraints(THD *thd, TABLE *table, const LEX_CSTRING *db_name,
   return schema_table_store_record(thd, table);
 }
 
+static int get_check_constraints_record(THD *thd, TABLE_LIST *tables,
+                                        TABLE *table, bool res,
```
[the remark this answers is not in the document]
Comment (an3l) [reply sql/sql_show.cc]: No it can't.

File: mysql-test/suite/funcs_1/t/is_check_constraints.test [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+
+set check_constraint_checks=1;
+
+use test;
+create table t0
+(
+ t int, check (t>32) # table constraint
+) ENGINE=myisam;
+
+--vertical_results
```
Comment (cvicentiu) [inline mysql-test/suite/funcs_1/t/is_check_constraints.test]: Please remove the vertical_results and horizontal_results modifiers. It doesn't improve readability in this case and it makes it harder to track.

File: mysql-test/suite/funcs_1/t/is_check_constraints.test [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+
+# Create table with same field and table check constraint name
+create table t3
+(
+a int,
+b int check (b>0), # field constraint named 'b'
+CONSTRAINT b check (b>10) # table constraint
+) ENGINE=InnoDB;
+
+--horizontal_results
```
Comment (cvicentiu) [inline mysql-test/suite/funcs_1/t/is_check_constraints.test]: Remove this line.

File: mysql-test/suite/funcs_1/t/is_check_constraints.test [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+select * from information_schema.check_constraints;
+
```
Comment (cvicentiu) [inline mysql-test/suite/funcs_1/t/is_check_constraints.test]: Please test creating an extra database and adding a table constraint there.

File: sql/sql_show.cc [hunk 11]
```diff
@@ -6819,6 +6819,42 @@ store_constraints(THD *thd, TABLE *table, const LEX_CSTRING *db_name,
...
+    if (thd->is_error())
+      push_warning(thd, Sql_condition::WARN_LEVEL_WARN,
+                   thd->get_stmt_da()->sql_errno(),
+                   thd->get_stmt_da()->message());
+    thd->clear_error();
+    DBUG_RETURN(0);
+  }
+  else if (!tables->view)
+  {
+    if (tables->table->s->table_check_constraints)
```
Comment (cvicentiu) [inline sql/sql_show.cc]: The if statement is unnecessary. The for loop will not execute if table_check_constraints is 0. Safe to delete.

File: sql/sql_show.cc [hunk 12]
```diff
@@ -6819,6 +6819,42 @@ store_constraints(THD *thd, TABLE *table, const LEX_CSTRING *db_name,
...
+    {
+      for (uint i= 0; i < tables->table->s->table_check_constraints; i++)
+      {
+        StringBuffer<MAX_FIELD_WIDTH> str(system_charset_info);
```
Comment (cvicentiu) [inline sql/sql_show.cc]: We are recreating the StringBuffer object on every for loop iteration. Although compiler might be smart enough to optimize this, let's do it ourselves. Move the object definition outside the for loop and just set the string's lenth to 0 before printing to it. When calling `check->print(&str);` make it:
```
str.length(0);
check->print(&str);
```

File: sql/sql_show.cc [hunk 13]
```diff
@@ -6819,6 +6819,42 @@ store_constraints(THD *thd, TABLE *table, const LEX_CSTRING *db_name,
...
+        Virtual_column_info *check= tables->table->check_constraints[i];
+        restore_record(table, s->default_values);
```
Comment (cvicentiu) [inline sql/sql_show.cc]: get_all_tables will pre-set table->field[idx_field1] and table->field[idx_field2] to be db name and table name. This means that we can optimize this like so:
Don't call restore_record at all.
Don't populate db_name and table_name at all.

Only store `def`, `check->name.str` and `str` from the printed check call.

File: sql/sql_show.cc [hunk 14]
```diff
@@ -6819,6 +6819,42 @@ store_constraints(THD *thd, TABLE *table, const LEX_CSTRING *db_name,
...
+        table->field[0]->store(STRING_WITH_LEN("def"), system_charset_info);
+        table->field[1]->store(db_name->str, db_name->length, system_charset_info);
+        table->field[2]->store(check->name.str, check->name.length, system_charset_info);
+        table->field[3]->store(table_name->str, table_name->length, system_charset_info);
+        check->print(&str);
+        table->field[4]->store(str.ptr(), str.length(), system_charset_info);
+        schema_table_store_record(thd, table);
```
Comment (cvicentiu) [inline sql/sql_show.cc]: You need to check the result of this call, it may fail. Return 1 if we failed.

File: sql/sql_show.cc [hunk 15]
```diff
@@ -9741,6 +9777,18 @@ ST_FIELD_INFO spatial_ref_sys_fields_info[]=
...
 
 
+ST_FIELD_INFO check_constraints_fields_info[]=
+{
+  {"CONSTRAINT_CATALOG", FN_REFLEN, MYSQL_TYPE_STRING, 0, 0, 0, OPEN_FULL_TABLE},
+  {"CONSTRAINT_SCHEMA", NAME_CHAR_LEN, MYSQL_TYPE_STRING, 0, 0, 0,
+   OPEN_FULL_TABLE},
+  {"CONSTRAINT_NAME", NAME_CHAR_LEN, MYSQL_TYPE_STRING, 0, 0, 0,
+   OPEN_FULL_TABLE},
+  {"TABLE_NAME", NAME_CHAR_LEN, MYSQL_TYPE_STRING, 0, 0, 0, OPEN_FULL_TABLE},
```
Comment (cvicentiu) [inline sql/sql_show.cc]: Let's move TABLE_NAME before CONSTRAINT_NAME. I think that makes more sense.

File: sql/sql_show.cc [hunk 16]
```diff
@@ -9756,6 +9804,8 @@ ST_SCHEMA_TABLE schema_tables[]=
    fill_schema_applicable_roles, 0, 0, -1, -1, 0, 0},
   {"CHARACTER_SETS", charsets_fields_info, 0,
    fill_schema_charsets, make_character_sets_old_format, 0, -1, -1, 0, 0},
+  {"CHECK_CONSTRAINTS", check_constraints_fields_info, 0,
+   get_all_tables, 0, get_check_constraints_record, 1, 2, 0, OPTIMIZE_I_S_TABLE|OPEN_TABLE_ONLY},
```
Comment (cvicentiu) [inline sql/sql_show.cc]: idx_field1 and idx_field2 are ok to be 1 and 2 if the order of the fields is CONSTRAINT_SCHEMA, TABLE_NAME.

File: sql/sql_show.cc (hunk 12 above)
Comment (cvicentiu) [reply sql/sql_show.cc]: Also, I see that you've changed this to system_charset_info. Looking at fb67cde2370f printing vcol was changed from using system_charset_info to my_charset_utf8mb4_general_ci. One needs to be able to answer the question as to why that was done there and why are we ok to keep system_charset_info here?

File: sql/sql_show.cc (hunk 16 above)
Comment (cvicentiu) [reply sql/sql_show.cc]: Also, respect 80 character rule please.

**Comment by cvicentiu** [comment]
Nicely done @an3l 

This is an excellent example of a *very good* pull request. It highlights all the aspects that one has to take care of when pushing a feature:

1. Write clear, clean code.
2. Test the code. Add comments in the tests where necessary.
3. Update other relevant tests, if any.

It also passes Travis tests now too. :)

Merged with:
d526679efd108478cc2af07578a15434fb46ed85

**Pull request closed without merging**